platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+{
+	int ret;
+	char *driver, *iommu_group;
+	char sysfs_path[2048], sysfs_link[2048];
+	DIR *dir;
+	struct dirent *dp;
+
+	/* Don't put / on the end of the path */
+	snprintf(sysfs_path, sizeof(sysfs_path),
+		 "/sys/class/net/%s/device/driver", if_name);
+	ret = mdev_readlink(sysfs_path, sysfs_link, sizeof(sysfs_link));
+	if (ret) {
+		ODP_ERR("Can't locate sysfs driver path\n");
+		return -1;
+	}
+
+	driver = mdev_basename(sysfs_link);
+	if (!driver) {
+		ODP_ERR("Can't driver in sysfs\n");
+		return -1;
+	}
+
+	if (strcmp(driver, mod_name)) {
+		ODP_ERR("Invalid driver name\n");
+		return -1;
+	}
+
+	snprintf(sysfs_path, sizeof(sysfs_path),
+		 "/sys/class/net/%s/device/mdev_supported_types/%s-netmdev/devices/",
+		 if_name, driver);
+
+	dir = opendir(sysfs_path);
+	if (!dir)
+		return -1;
+	/* FIXME only the last uuid will be returned now */
+	while ((dp = readdir(dir))) {
+		if (!strcmp(dp->d_name, ".") || !strcmp(dp->d_name, ".."))
+			continue;
+		else
+			strncpy(uuid, dp->d_name, sz);
+	}
+	closedir(dir);
+
+	if (!uuid[0])
+		return -1;
+
+	snprintf(sysfs_path, sizeof(sysfs_path),
+		 "/sys/bus/mdev/devices/%s/iommu_group", uuid);
+	ret = mdev_readlink(sysfs_path, sysfs_link, sizeof(sysfs_link));
+	if (ret) {
+		ODP_ERR("Can't locate IOMMU sysfs path\n");
+		return -1;
+	}
+
+	iommu_group = mdev_basename(sysfs_link);
+	if (!iommu_group) {
+		ODP_ERR("Can't locate iommu group in sysfs\n");
+		return -1;
+	}
+	ret = atoi(iommu_group);
+
+	return ret;
+}
+
+/**
+ * Initialize VFIO variables.
+ * set IOMMU and get device regions
+ */
+static int vfio_init_dev(int grp, int container,
+			 struct vfio_group_status *grp_status,
+			 struct vfio_iommu_type1_info *iommu_info,
+			 struct vfio_device_info *dev_info, char *grp_uuid)
+{
+	int device = -1;
+	int ret;
+
+	ret = ioctl(container, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU);
+	if (ret < 0) {
+		ODP_ERR("Doesn't support the IOMMU driver we want\n");
+		goto out;
+	}
+
+	/* Test the group is viable and available */
+	ret = ioctl(grp, VFIO_GROUP_GET_STATUS, grp_status);
+	if (ret < 0 || !(grp_status->flags & VFIO_GROUP_FLAGS_VIABLE)) {
+		ODP_ERR("Can't get status\n");
+		goto out;
+	}
+
+	ret = ioctl(grp, VFIO_GROUP_SET_CONTAINER, &container);
+	if (ret < 0) {
+		ODP_ERR("Failed to set container\n");
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+	if (ret < 0) {
+		ODP_ERR("Failed to set IOMMU\n");
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_IOMMU_GET_INFO, iommu_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get IOMMU info\n");
+		goto out;
+	}
+
+	ODP_DBG("iova_pgsizes bitmask=0x%llx\n", iommu_info->iova_pgsizes);
+	/* Get a file descriptor for the device */
+	device = ioctl(grp, VFIO_GROUP_GET_DEVICE_FD, grp_uuid);
+	if (device < 0) {
+		ODP_ERR("Failed to get device FD\n");
+		goto out;
+	}
+
+	/* Test and setup the device */
+	ret = ioctl(device, VFIO_DEVICE_GET_INFO, dev_info);
+	if (ret < 0) {
+		ODP_ERR("Failed to get device info\n");
+		goto out;
```

Comment:
`device` is not reset here so it will still be returned if this test fails. Is that intentional?

Reply:
Thanks, fixed.